smolyay/grid.py

```diff
@@ -0,0 +1,362 @@
+    ``dimension`` represents the number of independent variables of
+    the function to be approximated.
+
+    Through Smolyak technique of sampling, each dimension is assigned
+    an indicy which is integers from 1 to approximation level + 1
+    (The approximation level is called "exactness", and the length of
+    the levels, generated by :class:`NestedBasisSet`,
+    represents the exactness).
+
+    ..math::
+        n: dimensionality
+        \mu: exactness
+        [k_1, k_2, ..., k_n] = Smolyak indices
+        where k_1,2,...,n = [1, 2, ..., exactness +1]
+        n <= k_1 + k_2 + ... + k_n <= n + \mu
+
+    For instance:
+    ..math::
+        if n = 2 and \mu = 1, then:
+            if \sum_{1}^{n} k_i = 2:
+                (k_1 = 1, k_2 = 1)
+            if \sum_{i=1}^{n} k_i = 3:
+                (k_1 = 1, k_2 = 2)
+                (k_1 = 2, k_2 = 1)
+
+    :meth:`__call__`  generates the indexes of
+    grid points (integer grids) depending on dimensionality and
+    exactness. First, as above, all possible combinations of
+    indices are computed, which then can be replaced with indexes based on the
+    levels. Then, grid points are generated by making all combinations of
+    indexes for each combination of levels' indices.
+
+    Once integer grid points are generated, actual grid points
+    (``grid_points``) and their corresponding basis functions
+    (``grid_points_basis``) can be made by simply replacing
+    the indexes with their corresponding points and basis functions.
+    All the mentioned outputs are stored in an :class:`NestedIndexGrid`
+    object.
+
+
+    Parameters
+    ----------
+    nested_basis_set: :class:`NestedBasisSet`
+        points, basis functions and levels.
+
+    Example
+    -------
+    For instance, consider dimension (or n) = 2, and
+    exactness (or \mu) = 1 and Chebyshev's polynomial
+    of first kind, and its extremums as the basis function and points, then:
+    ..math::
+        levels = [[0], [1,2]]
+        level compositions = [([0],[0]), ([0],[1,2]), ([1,2],[0])]
+        then:
+        grid points indexes = [(0,0), (0,1), (0,2), (1,0), (2,0)]
+        points = [0, -1, 1]
+        basis = [:class:`ChebyshevFunction`, :class:`ChebyshevFunction`,
+                 :class:`ChebyshevFunction`]
+
+    Then:
+    ..math:
+        grid_points = [(0, 0), (0, -1), (0, 1), (-1, 0), (1, 0)]
+        grid points' functions = [(:class:`ChebyshevFunction`,
+        :class:`ChebyshevFunction`)
+        , (:class:`ChebyshevFunction`, :class:`ChebyshevFunction`),
+        (:class:`ChebyshevFunction`, :class:`ChebyshevFunction`),
+        (:class:`ChebyshevFunction`, :class:`ChebyshevFunction`),
+        (:class:`ChebyshevFunction`, :class:`ChebyshevFunction`)]
+
+    """
+
+    def __init__(self, nested_basis_set):
+        super().__init__(nested_basis_set)
+        self._nested_basis_set = nested_basis_set
+        self._levels = self._nested_basis_set.levels
```

Comment:
The levels will be in the nested basis set, so you don't need it. Also, use the same notation as the parent constructor, but document the required type of the argument and error check it:
```suggestion
    def __init__(self, basis_set):
        super().__init__(basis_set)
        if not isinstance(self.basis_set, NestedBasisSet):
            raise TypeError('Basis set must be nested for Smolyak grid')
```